packages-exp/auth-exp/src/core/user/user_impl.test.ts

```diff
@@ -0,0 +1,108 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { expect, use } from 'chai';
+import * as chaiAsPromised from 'chai-as-promised';
+import { UserImpl } from './user_impl';
+import { mockAuth } from '../../../test/mock_auth';
+import { StsTokenManager } from './token_manager';
+import { IdTokenResponse } from '../../model/id_token';
+
+use(chaiAsPromised);
+
+describe('core/user/user_impl', () => {
+  const auth = mockAuth('foo', 'i-am-the-api-key');
+  let stsTokenManager: StsTokenManager;
+
+  beforeEach(() => {
+    stsTokenManager = new StsTokenManager();
+  });
+
+  describe('constructor', () => {
+    it('attaches required fields', () => {
+      const user = new UserImpl({ uid: 'uid', auth, stsTokenManager });
+      expect(user.auth).to.eq(auth);
+      expect(user.uid).to.eq('uid');
+    });
+
+    it('attaches optional fields if provided', () => {
+      const user = new UserImpl({
+        uid: 'uid',
+        auth,
+        stsTokenManager,
+        displayName: 'displayName',
+        email: 'email',
+        phoneNumber: 'phoneNumber',
+        photoURL: 'photoURL'
+      });
+
+      expect(user.displayName).to.eq('displayName');
+      expect(user.email).to.eq('email');
+      expect(user.phoneNumber).to.eq('phoneNumber');
+      expect(user.photoURL).to.eq('photoURL');
+    });
+
+    it('sets optional fields to null if not provided', () => {
+      const user = new UserImpl({ uid: 'uid', auth, stsTokenManager });
+      expect(user.displayName).to.eq(null);
+      expect(user.email).to.eq(null);
+      expect(user.phoneNumber).to.eq(null);
+      expect(user.photoURL).to.eq(null);
+    });
+  });
+
+  describe('#getIdToken', () => {
+    it('returns the raw token if refresh tokens are in order', async () => {
+      stsTokenManager.updateFromServerResponse({
+        idToken: 'id-token-string',
+        refreshToken: 'refresh-token-string',
+        expiresIn: '100000'
+      } as IdTokenResponse);
+
+      const user = new UserImpl({ uid: 'uid', auth, stsTokenManager });
+      const token = await user.getIdToken();
+      expect(token).to.eq('id-token-string');
+      expect(user.refreshToken).to.eq('refresh-token-string');
+    });
+
+    it('throws if refresh is required', async () => {
+      const user = new UserImpl({ uid: 'uid', auth, stsTokenManager });
+      await expect(user.getIdToken()).to.be.rejectedWith(Error);
```

Comment:
can you add the error type & message expected so it's not just a generic error? that way we know it's failing as expected instead of due to some other issue

Reply:
It's only a "temporary" error because it's a method stub. When I actually implement those there'll be actual tests for the legit errors that may occur. Do you still want the test to check for the "not implemented" error specifically?

Reply:
oh yeah, I guess you're right, I thought this was was an actual expected throw.  Up to you whether we check for the 'not implemented' string, IMO it would be easier to grep through tests to see which ones need updating, but it's a nit at best